jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/Unit.java

```diff
@@ -0,0 +1,212 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.deploy.providers;
+
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.eclipse.jetty.deploy.App;
+import org.eclipse.jetty.util.StringUtil;
+
+/**
+ * A Unit of deployment, a basename and all the associated
+ * paths that belong to that basename, along with state.
+ */
+public class Unit
+{
+    public enum State
+    {
+        UNCHANGED,
+        ADDED,
+        CHANGED,
+        REMOVED
+    }
+
+    private final String baseName;
+    private final Map<Path, State> paths = new HashMap<>();
+    /**
+     * Indicates a synthetic Unit, used to track Environment specific configurations.
+     * Such as {@code <env>.xml}, {@code <env>.properties}, {@code <env>-<desc>.xml}, {@code <env>-<desc>.properties},
+     * all under a single unit that will not actually be deployed.
+     */
+    private String envConfig;
+    private App app;
+    private State state = State.UNCHANGED;
+
+    public Unit(String basename)
+    {
+        this.baseName = basename;
+    }
+
+    public String getBaseName()
+    {
+        return baseName;
+    }
+
+    public State getState()
+    {
+        return state;
+    }
+
+    private State calcState()
+    {
+        if (paths.isEmpty())
+            return State.REMOVED;
+
+        // Calculate state of unit from Path states.
+        Unit.State ret = null;
+        for (Unit.State pathState : paths.values())
```

Comment:
The algorithm here needs to be well documented in a comment so we can check if it is correct.     A unit is:

 + REMOVED if it is empty or all its deployable paths (xml, war, directory) are REMOVED
 + CHANGED if it has at least one path that is not UNCHANGED. If that one path is ADDED, then it must have another path that is not ADDED
 + ADDED if it has one or more deployable paths that are ADDED and none that are not ADDED

Note the inclusion of the concept of a deployable path (xml, war, directory), as only changes to them may result in ADDED or REMOVED.  However a change to a non deployable path (properties or .d) can result in a CHANGED.

Reply:
I don't think the implementation is right (also I think my description above might need improvement as well).

Reply:
I added some new javadoc, and some tests of `Unit.calcState()`, and fixed the implementation some to address these concerns better.

https://github.com/jetty/jetty.project/blob/fix/12.1.x/unify-deploy/jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/providers/UnitTest.java

Reply:
I still think it is not right.